File: 자료구조/C++/백준-스택-1725.cpp
#include <iostream>
#include <algorithm>
#include <stack>
#include <vector>
using namespace std;

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    int N;
    cin >> N;
    vector<int> h(N + 2);

    stack<int> s;
    s.push(0);

    int result = 0;

    for (int i = 1; i <= N; i++)
    {
        cin >> h[i];
    }

    for (int i = 1; i <= N + 1; i++)
    {
        while (!s.empty() && h[s.top()] > h[i])
        {
            int check = s.top();
            s.pop();
            result = max(result, h[check] * (i - s.top() - 1));
        }
        s.push(i);
    }

    cout << result << endl;
    return 0;
}

File: 자료구조/Python/백준-스택-1725.py
from sys import stdin

def main():
    N = int(stdin.readline())
    height = [int(stdin.readline()) for _ in range(N)]
    stack = []
    result = 0

    height.append(0)
    stack.append((0, height[0]))

    for i in range(1, N+1):
        left = i
        while stack and stack[-1][1] > height[i]:
            left, h = stack.pop()
            result = max(result, (i-left)*h)
        stack.append((left, height[i]))

    print(result)

if __name__ == "__main__":
    main()